File: Code Forces/Practice/Graph/party.cpp
/*
                C++
                encoding:UTF-8
                Modified: <25/Feb/2019 11:47:19 PM>

                ✪ H4WK3yE乡
                Mohd. Farhan Tahir
                Indian Institute Of Information Technology (IIIT),Gwalior
*/

#include <bits/stdc++.h>
using namespace std;

#define ll long long
#define mod 1000000007
#define sum(v) accumulate(v.begin(), v.end(), 0);
#define reset(a, b) memset(a, b, sizeof(a));
#define pb push_back

const int N = 1e5 + 5;
int n, scc, visited[N], component[N], check[N], belong[N];
vector<int> graph[N];

void dfs(int node) {
  visited[node] = 1;
  component[scc]++;
  belong[node] = scc;
  for (int neighbour : graph[node]) {
    if (!visited[neighbour])
      dfs(neighbour);
  }
}

int main() {
  ios_base::sync_with_stdio(false), cin.tie(0), cout.tie(0);
  cin >> n;
  int k;
  cin >> k;
  int u, v;
  for (int i = 0; i < k; i++) {
    cin >> u >> v;
    graph[u].pb(v);
    graph[v].pb(u);
  }
  for (int i = 1; i <= n; i++) {
    if (!visited[i]) {
      scc++;
      dfs(i);
    }
  }
  cin >> k;
  for (int i = 0; i < k; i++) {
    cin >> u >> v;
    if (belong[u] == belong[v]) {
      component[belong[u]] = 0;
    }
  }
  int mx = 0;
  for (int i = 1; i <= scc; i++) {
    mx = max(mx, component[i]);
  }
  cout << mx << "\n";

  return 0;
}

File: Code Forces/Practice/DFS/party.py
#	!/usr/bin/env python3
#	coding: UTF-8
#	Modified: <03/Feb/2019 01:19:28 PM>


#	✪ H4WK3yE乡
#	Mohd. Farhan Tahir
#	Indian Institute Of Information Technology (IIIT),Gwalior

#	Question Link
#
#

# ///==========Libraries, Constants and Functions=============///


import sys
sys.setrecursionlimit(10**6)
inf = float("inf")
mod = 1000000007


def get_array(): return list(map(int, sys.stdin.readline().split()))


def get_ints(): return map(int, sys.stdin.readline().split())


def input(): return sys.stdin.readline()

# ///==========MAIN=============///


visited = [False] * 2005
dist = [1] * 2005
graph = [[] for i in range(2005)]
root = []


def explore(node):
    visited[node] = True
    for child in graph[node]:
        if visited[child] == False:
            dist[child] = 1 + dist[node]
            explore(child)


def dfs(n):
    for i in root:
        if visited[i] == False:
            explore(i)


def main():
    n = int(input())
    for i in range(1, n + 1):
        x = int(input())
        if x != -1:
            graph[x].append(i)
        else:
            root.append(i)
    dfs(n)
    print(max(dist[1:n + 1]))


if __name__ == "__main__":
    main()
